### ISSUE TITLE
Dynamically link rocksdb during development

### ISSUE THREAD

**Issue by brson** [tikv/tikv#4338]
RocksDB is likely a big part of the TiKV binary. Linking to it dynamically during testing would reduce static linking time. We especially need to be able to do this on the release build machines, to reduce benchmarking time.

Related to https://github.com/tikv/tikv/issues/4151

**Comment by yiwu-arbug** [comment]
May be a dump question. How do we handle versioning of dynamic lib? With RocksDB we cherry-pick commits (bug fixes, new APIs) from upstream from time to time and don't have fixed release cycle.

**Comment by siddontang** [comment]
@yiwu-arbug 

Seem TiKV depends on specified version of rust-rocksdb, we can build the RocksDB in rust-rocksdb and put all the so files in `/bin/tikv` beside `tikv-server`, maybe we should explicitly set dynamic export path and let tikv-server use RocksDB now :-) 

But we should know that if we change RocksDB, we still need to rebuild all. Another think is for deployment, now we not only need to depoly the binary tikv-server, but also need to deploy all associated dynamic libs to the target machines.

**Comment by BusyJay** [comment]
I think we can export a tar instead of a binary so that all required library can be shipped and deployed automatically by ansible or k8s.

**Comment by brson** [comment]
@yiwu-arbug I don't think it matters very much. These dynamic libraries are only going to be used for development purposes, not distributed.

I created a branch of rust-rocksdb here: https://github.com/brson/rust-rocksdb/compare/master...brson:dylib?expand=1

It doesn't link yet, but I am hopeful that with a single change to rocksdb itself that it will. bzip2, as configured w/o stdio, requires clients to define an error handler called bz_internal_error. It seems that when static linked to tikv this isn't required, probably because the linker throws that symbol out as unused. But the dynamic rocksdb build fails during linking looking for that symbol.

So I'm hopeful that by defining that symbol tikv will build and link at runtime.

That said, the proper fix to enable linking rocksdb as a dylib may be more involved. And it has to do with the 4 compression libraries we link rocks to. Those are pulled into librocksdb_sys by making the bzip2, lz4 crates, etc dependencies. Then during a static build those get linked in, along with rocksdb. With the dylib build though, rocksdb's build system statically links the compression libraries into itself; but also they are going to be linked by rustc into the final exes too.

For the symbols that don't get automatically removed from the exe, I don't know what the dynamic linker will do to resolve the duplicate functions.

Upstream librocksdb_sys builds all the compression libraries itself, which may avoid some problems here.

**Comment by brson** [comment]
Difining `bz_internal_error` in rocksdb itself allows the static linking to complete, and seems to produce a working tikv-server bin.

It would be better to define `bz_internal_error` elsewhere though.

**Comment by yiwu-arbug** [comment]
@siddontang I missed that we always bump rust-rocksdb version whenever we update rocksdb. Thanks.

**Comment by brson** [comment]
I don't see a significant difference in compile times by dynamic linking rocks:

```
`time cargo build --release`

real    26m11.710s
user    54m29.625s
sys     1m58.802s

real    26m8.029s
user    54m19.496s
sys     1m58.591s

dynamic rocks:

`time cargo build --release --features=shared-rocksdb`

real    26m5.035s
user    54m14.631s
sys     1m54.971s

real    26m1.859s
user    54m14.992s
sys     1m54.576s
```

I verified that the static case was statically linked and the dynamic case was dynamically linked.

**Comment by siddontang** [comment]
Very bad news...

**Comment by brson** [comment]
Here are my present branches:

This is the rust-rocksdb branch: https://github.com/brson/rust-rocksdb/tree/dylib

The rocksdb submodule must have this branch fetched: https://github.com/brson/rocksdb/tree/dylib

This is the tikv branch: https://github.com/brson/tikv/tree/dylib

It expects rust-rocksdb to be checked out to ../rust-rocksdb. Build tikv with --features=shared-rocksdb

**Comment by siddontang** [comment]
I do the same experiments and get the same result.
@BusyJay @breeswish can you try it again?
